### PR TITLE
add py3k patch

### PR DESCRIPTION
**Pull request by kmuehlbauer**
This adds (hopefully) py3k support to eccodes to at least linux. MacOSX should work, but this needs to be proved. The patch was created locally.

Reading and writing data files is working. I tested the examples locally on linux. I didn't get the testsuit to work, although they run locally by hand.

There are some glitches left, mostly related to binary/string problems with py3k when handling raw binary messages, but for most use cases this will do.

The necessary tweaks for the swig intermediate file were applied from https://github.com/pygraphviz/pygraphviz

### CODE REVIEW

**Comment by ocefpaf** [comment]
@kmuehlbauer you need to rerender with conda smithy to add Python 3 to the build matrix:

```shell
conda install -c conda-forge conda-smithy  # Must be your root env.
conda smithy rerender  # From the feedstock directory.
```

then commit the changes.

File: .travis.yml [hunk 1]
```diff
@@ -8,6 +8,10 @@ env:
     
     - CONDA_NPY=110  CONDA_PERL=5.20.3.1  CONDA_PY=27
     - CONDA_NPY=111  CONDA_PERL=5.20.3.1  CONDA_PY=27
+    - CONDA_NPY=110  CONDA_PERL=5.20.3.1  CONDA_PY=34
+    - CONDA_NPY=111  CONDA_PERL=5.20.3.1  CONDA_PY=34
+    - CONDA_NPY=110  CONDA_PERL=5.20.3.1  CONDA_PY=35
+    - CONDA_NPY=111  CONDA_PERL=5.20.3.1  CONDA_PY=35
```
Comment (ocefpaf) [inline .travis.yml]: :+1:

**Comment by kmuehlbauer** [comment]
@ocefpaf There is trouble ahead:

Packages missing in current linux-64 channels: 
  - eccodes 0.13.0
  - jasper

**Comment by ocefpaf** [comment]
> @ocefpaf There is trouble ahead

It seems to be working on `CircleCI`and `Travis-CI`. Did you see that failure locally? `conda` just had a few updates that broke and fixed the channel addition, so check your channels.

**Comment by kmuehlbauer** [comment]
`conda-smithy` was complaining about that, but went through.

But, as it turns out: It works also on OSX. And at the first attempt. 

I can check the linux stuff if that gets merged.

**Comment by kmuehlbauer** [comment]
:tada: Green lights. :green_heart:

**Comment by kmuehlbauer** [comment]
> Edit: check the .condarc file and see if conda-forge and defaults are there.

Yes, this is `.condarc`:

```
channels:
  - conda-forge
  - defaults
```

**Comment by kmuehlbauer** [comment]
As these changes are nearly all about `gribapi`, we can try to have (part) of that patch with @conda-forge/python-ecmwf_grib.

**Comment by ocefpaf** [comment]
> As these changes are nearly all about `gribapi`, we can try to have (part) of that patch with @conda-forge/python-ecmwf_grib. 

BTW we had a patch for py3k compatibility in `python-ecmwf_grib`, created by @Qulogic, that we dropped. I don't recall the reason.

**Comment by ocefpaf** [comment]
@kmuehlbauer ready to merge?

**Comment by kmuehlbauer** [comment]
> BTW we had a patch for py3k compatibility in python-ecmwf_grib, created by @Qulogic, that we dropped. I don't recall the reason.

Do you have a pointer?

**Comment by ocefpaf** [comment]
> Do you have a pointer?

Found it!

https://github.com/SciTools/conda-recipes-scitools/issues/88

We need to check if `eccodes` is not affect by the same issue.

**Comment by kmuehlbauer** [comment]
> Found it!
> 
> SciTools/conda-recipes-scitools#88
> 
> We need to check if eccodes is not affect by the same issue.

Sure, checking is necessary. But this here uses new ecmwf-`cmake` build. I'm trying to get `ecmwf_grib` patched also.

**Comment by kmuehlbauer** [comment]
@ocefpaf You can merge this, ready from my side.

**Comment by pelson** [comment]
I think we need to be super careful here. I agree with the desire to py3k eccodes, but it isn't and shouldn't be a conda-forge patch to achieve it. Is there a related issue on the eccodes issue tracker (if so, please add it as a comment to the recipe)? Are the ECMWF aware of this patch?

In many ways, if the lack of a public repo for eccodes makes it hard to submit a PR, it would be healthier to make a friendly fork on GitHub and use that as the canonical source here.

Just to be clear, I'm extremely pleased we can use py3k with eccodes, I just want to be careful how we do it and ensure that ownership lies with the code maintainers, not the packagers.

👍

**Comment by kmuehlbauer** [comment]
@pelson 

I have two issues open at the ecmwf issue tracker (on eccodes), which are not related to py3k. But I added a comment about this particular patch at one issue. 

I see your point about the canonical source here. I have imported `eccodes` under my account (https://github.com/kmuehlbauer/eccodes) where I have testing the py3k-patch for `eccodes` (. This could be done also with `gribapi`. I had some problems though, to import things correctly b/c `.gitignore` is somehow broken in the source tarball. 

If there is consensus, I can re-create the friendly fork with the whole source tarball and we can proceed from there. I can also create the friendly fork here (https://github.com/meteo-ubonn) at my institutes organization. Suggestions?

**Comment by pelson** [comment]
> If there is consensus, I can re-create the friendly fork with the whole source tarball and we can proceed from there. I can also create the friendly fork here (https://github.com/meteo-ubonn) at my institutes organization. Suggestions?

Are you aware of any ECMWF folk we could ping here? Making a friendly fork is not ideal, but may be necessary if we want to *add capabilities* to packaged software. The ideal solution is that the patch is accepted (at least) and released in a formal eccodes version, but it would be healthy to understand the realistic timescales of achieving that.

In short - it would be *extremely* helpful to get ECMWF involvement on this.

**Comment by kmuehlbauer** [comment]
> Are you aware of any ECMWF folk we could ping here?

Unfortunately no. Also getting access to the issue tracker is a pain. The response times are sometimes fairly good (1-2 days), but sometimes you are waiting some more days.

The most problematic thing is, that we do not know which way ecmwf is heading with the python bindings and the whole `eccodes` library itself. We do not know, if they will stick to the `FILE* model` or think about something else. From what I extracted from online resources (http://www.wmo.int/pages/prog/sat/meetings/documents/IPET-SUP-2_Doc_10-03_ecCodes-Update.pdf) the release of 2.0 is set for before end 2016, which isn't that far away.

I have some doubts about python 3 support in release 2.0. Their answer about that was, they have plans for that and registered my interest on that matter.

Anyway, what I can actually do is to open another issue on their issue tracker, asking about their thoughts on a friendly fork to speed up community testing, py3k-issues and more.

@pelson Is this the way to go?

**Comment by pelson** [comment]
> @pelson Is this the way to go?

I think that is a very healthy approach - the lack of an open development model of eccodes makes it hard to make an informed decision, and the sooner we can open the communication channels, the better IMO.

**Comment by kmuehlbauer** [comment]
I opened an issue about this at ecmwf and report back when there is something happening.

**Pull request merged**